### DISCUSSION TITLE
getSelectedElements typescript issue?

### DISCUSSION THREAD

**Discussion by AmosSzust** [bcakmakoglu/vue-flow#1229] [Q&A] [answered]
Hi

I did the following in order to delete a selected element
```
      if (!this.vueFlowInstance) return;
      const selected: ComputedRef<FlowElements> = this.vueFlowInstance.getSelectedElements;
      if (selected.value.length === 0) return;
      const index = this.flowNodes.indexOf(selected.value[0]);
      if (index !== -1) {
        this.flowNodes.splice(index, 1);
      }
```
Typescript accepts it while developing but in runtime, the data is actually in `selected` and `selected.value` is undefined causing the above to fail.

If I do just `selected.length === 0` obviously typescript doesn't accept it, telling me 

> Property 'length' does not exist on type 'ComputedRef<FlowElements>'

What am I missing?

**Comment by bcakmakoglu** [comment]
Can you provide a reproduction for this issue?

[You can find a sandbox template here.](https://codesandbox.io/p/devbox/vue-flow-basic-gfgro4)

I recommend you start using composition API instead of options API, though that is ultimately your choice of course but you'll make your life easier in the future 😄

**Reply by bcakmakoglu** [reply]
By composition API I mean this: https://vuejs.org/api/composition-api-setup.html#composition-api-setup

The problem you're encountering is that you're simply mistyping this here `const selected: ComputedRef<FlowElements> = this.vueFlowInstance.getSelectedElements;`

I assume your component looks something like this

```ts
export default defineComponent({
  name: 'BasicOptionsAPI',
  components: { VueFlow, Background, MiniMap, Controls },
  data() {
    return {
      instance: null as VueFlowStore | null,
      elements: [/* ... */] as Elements,
    }
  },
  methods: {
    onPaneReady(instance: FlowEvents['paneReady']) {
      instance.fitView()
      this.instance = instance as any
    },
    onConnect(params: FlowEvents['connect']) {
      this.instance?.addEdges(params)
    },
  },
})
```

You have to consider that `ref` is *auto-unwrapped* in the options API (though this might be harder to grasp if you're unfamiliar with the composition api in general), therefore you do not use `this.instance.someRef.value` but `this.instance.someRef`.

A short explanation of what `ref` is and what *auto-unwrapped* means.

In composition API, we define component methods, data etc. in the *setup function of the component*.
This setup function is run *once* when the component is created.

A basic component looks like this (using the *script setup* syntax)

```vue
<script setup>
import { ref } from 'vue'

// this is a `ref` - previously you would define this in `data` of a component using options API `export default { data() { msg: 'Hello World!' } }`
const msg = ref('Hello World!')

// Accessing a refs value is done using `ref.value` when used *in any context besides the template of a component*
console.log(msg.value)
</script>

<template>
  <!-- Note that we *do not* have to use `msg.value` in the template - it has been *auto-unwrapped* for us -->
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
```

You'll have to type-cast `instance as any` though.

**Reply by AmosSzust** [reply]
My implementation looks like this, in template section I have the following
```
    <VueFlow
      v-model="flowBlocks"
      style="height: calc(100vh - 110px)"
      fit-view-on-init
      auto-connect
      @pane-ready="paneReady"
    />

```
I have imported the following: `import { Elements, VueFlow, VueFlowStore, useVueFlow } from "@vue-flow/core";`
`components: { VueFlow },`
in the setup, I have returned the following
```
    return {
      useVueFlow, // though I'm not really using it anywhere
      ...
    }
```
and in paneReady event I did the following
```
    paneReady(vueFlowInstance: VueFlowStore) {
      this.vueFlowInstance = vueFlowInstance;
    },
```
and I added a button to delete the selected element by the user
```
    deleteSelected() {
      if (!this.vueFlowInstance) return;
      const selectedNodes = this.vueFlowInstance.getSelectedNodes.value;

      const selectedEdges = this.vueFlowInstance.getSelectedEdges.value;

      if (selectedNodes.length) {
        this.vueFlowInstance.removeNodes(selectedNodes);
      }

      if (selectedEdges.length) {
        this.vueFlowInstance.removeEdges(selectedEdges);
      }
    },
```
but as I mentioned, I guess I'm missing something because the .value is confusing me between design time and runtime.

**Reply by bcakmakoglu** [reply]
Yeah, as mentioned, take a dive into the composition API Vue docs, it'll help you get a better understanding of the differences here.

**Reply by AmosSzust** [reply]
I took a dive here https://vuejs.org/api/composition-api-setup and saw that I'm already using composition api and even shared some code. Not sure what part of my code suggests that I don't use composition api.

```
export default defineComponent({
  name: "FlowPage",
  components: { VueFlow },
  setup() {
    const store = appStore();
    const flowBlocks = ref<Elements>([
      {
        id: "1",
        label: "Start",
        type: "input",
        position: { x: 200, y: 10 },
      },
    ]);
    return {
      getBadgeColor,
      useVueFlow,
...

```
Back to my simple(?) question, I wrote what I did, based on your wonderful explanation and shared the problem I encountered. Is it possible to explain what did I do wrong? Here is the code again
```
    deleteSelected() {
      if (!this.vueFlowInstance) return;
      const selectedNodes = this.vueFlowInstance.getSelectedNodes.value;

      const selectedEdges = this.vueFlowInstance.getSelectedEdges.value;

      if (selectedNodes.length) {
        this.vueFlowInstance.removeNodes(selectedNodes);
      }

      if (selectedEdges.length) {
        this.vueFlowInstance.removeEdges(selectedEdges);
      }
    },
```
I'll take this line as an example: `const selectedNodes = this.vueFlowInstance.getSelectedNodes.value;`
getSelectedNodes type is `(property) getSelectedNodes: ComputedRef<GraphNode<any, any, string>[]>` so in order to get the value I added `.value;` which in design time looks legit (without it I get an error in vscode) but at run time, the `.value` returns undefined because the actual data is here `this.vueFlowInstance.getSelectedNodes`.

I understand that I'm missing something (silly?) and I understand that it is unrelated to composition API. My question is, what am I missing?
Thanks

**Reply by AmosSzust** [reply]
I did some more digging and changed things a bit but again I encountered the same problem
```
  setup() {
    const store = appStore();
    const flowBlocks = ref<Elements>([
      {
        id: "1",
        label: "Start",
        type: "input",
        position: { x: 200, y: 10 },
      },
    ]);
    let vueFlowInstance1 = ref<VueFlowStore>();
    function deleteSelected1() {
      if (!vueFlowInstance1.value) return;
      const selectedNodes = vueFlowInstance1.value.getSelectedNodes.value;
      const selectedEdges = vueFlowInstance1.value.getSelectedEdges.value;
      if (selectedNodes.length) {
        vueFlowInstance1.value.removeNodes(selectedNodes);
      }
      if (selectedEdges.length) {
        vueFlowInstance1.value.removeEdges(selectedEdges);
      }
    }
    function paneReady1(vueFlowInstance: VueFlowStore) {
      vueFlowInstance1.value = vueFlowInstance;
    }
    return {
      getBadgeColor,
      useVueFlow,
      deleteSelected1,
      paneReady1,
   }
}
```
The following 2 lines have the same issue as I wrote in my original question, with `.value` typescript accepts it but in runtime `.value` contains undefined.
```
const selectedNodes = vueFlowInstance1.value.getSelectedNodes.value;
const selectedEdges = vueFlowInstance1.value.getSelectedEdges.value;
```

**Reply by bcakmakoglu** [reply]
A lot of things from the code you previously showed suggests that you were *not* using composition API (aside from the fact that you literally asked me what composition API even is 😂) - mainly using `this`.
There's no `this` in a setup function.

Aside from that, if you're already in a setup function, use the instance straight from `useVueFlow`.

```ts
const { getSelectedNodes } = useVueFlow();

console.log(getSelectedNodes.value)
```

That should resolve your type issue.

**Reply by AmosSzust** [reply]
I guess I did a mix between the 2. Anyway, now I did it correctly and the problem is that both `getSelectedNodes `and `getSelectedEdges `are empty even when I select a node/connection line.
```
    function deleteSelected1() {
      const { getSelectedNodes, getSelectedEdges } = useVueFlow();
      const selectedNodes = getSelectedNodes.value;
      const selectedEdges = getSelectedEdges.value;
      if (selectedNodes.length) {
        useVueFlow().removeNodes(selectedNodes);
      }
      if (selectedEdges.length) {
        useVueFlow().removeEdges(selectedEdges);
      }
    }

```

**Reply by AmosSzust** [reply]
Here is my sandbox example, click the "to delete" node to make sure it is selected and press the test button at the top. Let me know what am I missing.
https://codesandbox.io/p/devbox/eloquent-golick-2grnlp?file=%2Fsrc%2FApp.vue%3A13%2C48

**Reply by bcakmakoglu** [reply, marked as answer]
General rule of thumb: Composable are not meant to be used in callbacks / functions.
There are exceptions (since composables are ultimately *just functions*) but `useVueFlow` is not one of them.
It's similar to React Hooks (but more flexible).

Call the composable *in the setup body* of your component (you can use any example on the vueflow docs as a reference to see how it's *correctly* used).

```vue
<script setup lang="ts">
const { getSelectedNodes, getSelectedEdges } = useVueFlow()

function deleteSelectedElements() {
	if (getSelectedNodes.value.length) { /* ... */ }
	// ...
}
```

Or if you want to use `defineComponent`:
```ts
export default defineComponent({
	setup() {
		const { getSelectedNodes, getSelectedEdges } = useVueFlow()

		function deleteSelectedElements() {
			if (getSelectedNodes.value.length) { /* ... */ }
			// ...
		}	
	}
})
```

Please don't misunderstand me or take this the wrong way but these are fundamentals of the Vue Composition API and it's got nothing to do with Vue Flow 😄 Sure I could start throwing errors for incorrect usage (which I will in the future so you can't keep calling `useVueFlow` in a context where it's invalid to call it to begin with) but atm it doesn't, which isn't super beginner friendly and expects you to know how composables are supposed to work.

To explain why using `useVueFlow` isn't gonna work in a callback.

When you call `useVueFlow` what happens under the hood is:
1. Vue Flow will check the current context if a state can be found (an id will be written to `currentInstance`, which usually is the component that called `useVueFlow`.
2. If no state can be found from the context and no id was explicitly passed to `useVueFlow`, a new state will be created (and a new id for this state created and, if possible, written to the context)

During your function execution of `deleteSelected1` *there is no instance to find the state from* - `getCurrentInstance()` (which is a Vue built-in function) will not return the component instance so we can check if a state is available so you just generate a new state constantly.

[You can confirm this yourself by using this modified sandbox.](https://codesandbox.io/p/devbox/bold-fog-flj4qm?file=%2Fsrc%2FApp.vue%3A26%2C12)

As you can see, everytime you press the *Test* button, you will receive an alert *with a new id*.
